pkg/archive/archive_test.go

```diff
@@ -1236,6 +1236,64 @@ func TestXGlobalNoParent(t *testing.T) {
 	assert.Check(t, errors.Is(err, os.ErrNotExist))
 }
 
+// TestImpliedDirectoryPermissions ensures that directories implied by paths in the tar file, but without their own
+// header entries are created recursively with the default permissions stored in ImpliedDirectoryPermisions. This test
+// also verifies that the permissions of explicit directories are respected.
+func TestImpliedDirectoryPermissions(t *testing.T) {
+	skip.If(t, runtime.GOOS == "windows", "skipping test that requires Unix permissions")
+
+	buf := &bytes.Buffer{}
+	headers := []tar.Header{{
+		Name: "deeply/nested/and/implied",
+	}, {
+		Name: "explicit/",
+		Mode: 0644,
+	}, {
+		Name: "explicit/permissions/",
+		Mode: 0600,
+	}, {
+		Name: "explicit/permissions/specified",
+		Mode: 0400,
+	}}
+
+	w := tar.NewWriter(buf)
+	for _, header := range headers {
+		err := w.WriteHeader(&header)
+		assert.NilError(t, err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "implied-perms-test")
+	assert.NilError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	err = Untar(buf, tmpDir, nil)
+	assert.NilError(t, err)
+
+	stat, err := os.Lstat(filepath.Join(tmpDir, "deeply"))
+	assert.Check(t, err == nil)
+	assert.Check(t, stat.Mode().Perm() == ImpliedDirectoryPermisions)
+
+	stat, err = os.Lstat(filepath.Join(tmpDir, "deeply/nested"))
+	assert.Check(t, err == nil)
+	assert.Check(t, stat.Mode().Perm() == ImpliedDirectoryPermisions)
+
+	stat, err = os.Lstat(filepath.Join(tmpDir, "deeply/nested/and"))
+	assert.Check(t, err == nil)
+	assert.Check(t, stat.Mode().Perm() == ImpliedDirectoryPermisions)
+
+	stat, err = os.Lstat(filepath.Join(tmpDir, "explicit"))
+	assert.Check(t, err == nil)
+	assert.Check(t, stat.Mode().Perm() == 0644)
+
+	stat, err = os.Lstat(filepath.Join(tmpDir, "explicit/permissions"))
+	assert.Check(t, err == nil)
+	assert.Check(t, stat.Mode().Perm() == 0600)
+
+	stat, err = os.Lstat(filepath.Join(tmpDir, "explicit/permissions/specified"))
+	assert.Check(t, err == nil)
+	assert.Check(t, stat.Mode().Perm() == 0400)
```

Comment:
This sure is getting repetitive. Extract into a helper function/closure?
```go
assertPerm := func(path string, perm int64) {
	t.Helper()
	stat, err = os.Lstat(filepath.Join(tmpDir, "path))
	assert.Check(t, err)
	assert.Check(t, is.Equal(stat.Mode().Perm(), perm))
}

assertPerm("deeply/nested", ImpliedDirectoryPermissions)
assertPerm("deeply/nested/and", ImpliedDirectoryPermissions)
// etc...
```

Reply:
Thanks!